File: src/main/java/epicsquid/superiorshields/setup/compat/BotaniaCompat.java
package epicsquid.superiorshields.setup.compat;

import com.tterrag.registrate.util.nullness.NonNullFunction;
import epicsquid.superiorshields.item.BotaniaShieldItem;
import epicsquid.superiorshields.item.SuperiorShieldItem;
import epicsquid.superiorshields.item.VanillaShieldItem;
import epicsquid.superiorshields.shield.IShieldType;
import net.minecraft.world.item.Item;
import net.minecraftforge.fml.ModList;

public class BotaniaCompat {

	private static boolean LOADED;

	public static void init() {
		LOADED = ModList.get().isLoaded("botania");
	}

	public static NonNullFunction<Item.Properties, SuperiorShieldItem<IShieldType>> makeBotaniaShieldOrDefault(IShieldType type) {
		return LOADED ? LoadedOnly.makeBotaniaShield(type) : props -> new VanillaShieldItem(props.durability(512), type);
	}

	public static class LoadedOnly {

		public static NonNullFunction<Item.Properties, SuperiorShieldItem<IShieldType>> makeBotaniaShield(IShieldType type) {
			return props -> new BotaniaShieldItem(props, type);
		}
	}
}

File: src/main/kotlin/dev/epicsquid/superiorshields/compat/BotaniaCompat.kt
package dev.epicsquid.superiorshields.compat

import dev.epicsquid.superiorshields.item.ElementiumSuperiorShieldItem
import dev.epicsquid.superiorshields.item.SuperiorShieldItem
import dev.epicsquid.superiorshields.shield.BotaniaManaSuperiorShield
import dev.epicsquid.superiorshields.shield.DurabilitySuperiorShield
import net.minecraft.world.item.Item.Properties
import net.minecraft.world.item.crafting.Ingredient
import net.minecraftforge.fml.ModList

object BotaniaCompat {

	private val loaded: Boolean by lazy { ModList.get().isLoaded("botania") }

	fun botaniaManaShieldBuilder(
		props: Properties,
		enchantmentValue: Int,
		type: DurabilitySuperiorShield,
		repairItem: () -> Ingredient?,
	): SuperiorShieldItem<*> =
		if (loaded) {
			LoadedOnly.botaniaManaShieldBuilder(props, enchantmentValue, type, repairItem)
		} else {
			SuperiorShieldItem(
				props = props,
				enchantmentValue = enchantmentValue,
				type = type,
				repairItem = repairItem
			)
		}

	fun elementiumShieldBuilder(
		props: Properties,
		enchantmentValue: Int,
		type: DurabilitySuperiorShield,
		repairItem: () -> Ingredient?,
	): SuperiorShieldItem<*> =
		if (loaded) {
			LoadedOnly.elementiumShieldBuilder(props, enchantmentValue, type, repairItem)
		} else {
			SuperiorShieldItem(
				props = props,
				enchantmentValue = enchantmentValue,
				type = type,
				repairItem = repairItem
			)
		}

	object LoadedOnly {
		fun botaniaManaShieldBuilder(
			props: Properties,
			enchantmentValue: Int,
			type: DurabilitySuperiorShield,
			repairItem: () -> Ingredient?,
		): SuperiorShieldItem<*> =
			SuperiorShieldItem(
				props = props,
				enchantmentValue = enchantmentValue,
				type = BotaniaManaSuperiorShield(
					type.config
				),
				repairItem = repairItem,
			)

		fun elementiumShieldBuilder(
			props: Properties,
			enchantmentValue: Int,
			type: DurabilitySuperiorShield,
			repairItem: () -> Ingredient?,
		): SuperiorShieldItem<*> =
			ElementiumSuperiorShieldItem(
				props = props,
				enchantmentValue = enchantmentValue,
				type = BotaniaManaSuperiorShield(
					type.config
				),
				repairItem = repairItem
			)

	}
}